executor/write.go

```diff
@@ -816,43 +814,47 @@ func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (Row, error
 			return nil, errors.Trace(err)
 		}
 	}
+	if !e.IgnoreErr && len(e.OnDuplicate) > 0 {
+		var err error
+		rows, err = e.batchUpdateDupRows(rows, e.OnDuplicate)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
 	for _, row := range rows {
 		// duplicate row will be marked as nil in batchMarkDupRows if
 		// IgnoreErr is true. For IgnoreErr is false, it is a protection.
 		if row == nil {
 			continue
 		}
-		if batchInsert && rowCount >= batchSize {
-			e.ctx.StmtCommit()
-			if err := e.ctx.NewTxn(); err != nil {
-				// We should return a special error for batch insert.
-				return nil, ErrBatchInsertFail.Gen("BatchInsert failed with error: %v", err)
-			}
-			txn = e.ctx.Txn()
-			rowCount = 0
-			if !sessVars.ImportingData {
-				writeBufs.BufStore = kv.NewBufferStore(txn, kv.TempTxnMemBufCap)
-			}
+		if err := e.checkBatchLimit(); err != nil {
+			return nil, errors.Trace(err)
 		}
 		if len(e.OnDuplicate) == 0 && !e.IgnoreErr {
-			txn.SetOption(kv.PresumeKeyNotExists, nil)
+			e.ctx.Txn().SetOption(kv.PresumeKeyNotExists, nil)
 		}
 		h, err := e.Table.AddRecord(e.ctx, row, false)
-		txn.DelOption(kv.PresumeKeyNotExists)
+		e.ctx.Txn().DelOption(kv.PresumeKeyNotExists)
 		if err == nil {
 			if !sessVars.ImportingData {
 				e.ctx.StmtAddDirtyTableOP(DirtyTableAddRow, e.Table.Meta().ID, h, row)
 			}
-			rowCount++
+			e.rowCount++
 			continue
 		}
 
 		if kv.ErrKeyExists.Equal(err) {
-			if len(e.OnDuplicate) > 0 {
-				if err = e.onDuplicateUpdate(row, h, e.OnDuplicate); err != nil {
+			if e.IgnoreErr && len(e.OnDuplicate) > 0 {
+				data, err1 := e.Table.RowWithCols(e.ctx, h, e.Table.WritableCols())
```

Comment:
This line sends Get operation to TiKV which can be avoided.

Reply:
It is used for `insert ignore on duplicate key update`, which is not the purpose for this optimization. I'll try to deal with it some days later.